monai/data/dataset.py

```diff
@@ -321,6 +323,238 @@ def __getitem__(self, index):
+                return True
+            else:
+                return False
+
+    def update_cache(self):
+        """
+        Update cache items for current epoch, need to call this function before every epoch.
+        If the cache has been shutdown before, need to restart the `_replace_mgr` thread.
+
+        """
+        if not self._replace_mgr.isAlive():
+            self._restart()
+
+        # make sure update is done
+        while True:
+            if self._try_update_cache():
+                break
+            else:
+                time.sleep(0.01)
+
+    def _try_shutdown(self):
+        """
+        Wait for thread lock to shut down the background thread.
+
+        """
+        with self._update_lock:
+            if self._round_done:
+                self._round = 0
+                self._round_done = False
+                return True
+            else:
+                return False
+
+    def shutdown(self):
+        """
+        Shut down the background thread for replacement.
+
+        """
+        if not self._started:
+            return
+
+        # wait until replace mgr is done the current round
+        while True:
+            if self._try_shutdown():
+                self._replace_mgr.join()
+                return
+            else:
+                time.sleep(0.01)
+
+    def _replace_cache_thread(self, index: int):
+        """
+        Execute deterministic transforms on the new data for replacement.
+
+        """
+        pos: int = self._replace_data_idx[index]
+        self._replacements[index] = self._load_cache_item(self.data[pos], self.transform.transforms)  # type: ignore
+
+    def _compute_replacements(self):
+        """
+        Compute expected items for the replacement of next epoch, execute deterministic transforms.
+        It can support multi-threads to accelerate the computation progress.
+
+        """
+        if self.num_replace_workers > 0:
+            with ThreadPool(self.num_replace_workers) as p:
+                p.map(self._replace_cache_thread, list(range(self._replace_num)))
+        else:
+            for i in range(self._replace_num):
+                pos: int = self._replace_data_idx[i]
+                self._replacements[i] = self._load_cache_item(self.data[pos], self.transform.transforms)  # type: ignore
+        self._round_done = True
+
+    def _try_manage_replacement(self, check_round):
+        """
+        Wait thread lock and replace training items in the bacground thread.
+
+        """
+        with self._update_lock:
+            if self._round <= 0:
+                # shutdown replacement
+                self._round_done = True
+                return True, -1
+
+            if self._round != check_round:
+                self._compute_replacements()
+            return False, self._round
+
+    def manage_replacement(self):
+        """
+        Background thread for replacement.
+
+        """
+        check_round: int = -1
+        while True:
```

Comment:
I generally recommend against using infinite loops like this since it's easier to read and better practice to have the actual condition the loop relies on to be in the condition expression itself. If ever we use actual formal analysis with this code it can be harder to reason about this sort of logic too. I'd put `done = False` above this line then have `while not done:` as this line, then lines 546-7 aren't needed. I mentioned the issue above with daemon threads, this loop might still have problems with not terminating if the thread isn't a deamon.

Reply:
OK, updated all the `while True` logic in the latest commit.
Thanks.